include/lsst/afw/image/detail/MaskDict.h

```diff
@@ -0,0 +1,185 @@
+ * automatically added to *all* active Mask instances that do not already use
+ * that bit (via MaskDict::addMaskPlane) as well as a default MaskDict that is
+ * used when constructing future Mask instances.  In contrast, when a plane is
+ * removed, it typically only affects that Mask, and if that Mask is currently
+ * using the default MaskDict, the default is redefined to a copy prior to
+ * the plane's removal (via detachDefault).
+ *
+ * To maintain that global state, all MaskDicts must be held by shared_ptr,
+ * and hence MaskDict's constructors are private, and static or instance
+ * methods that return shared_ptr are provided to replace them.
+ *
+ * MaskDict is an implementation detail (albeit and important one) and hence
+ * its "documentation" is intentionally in the form of regular comments
+ * rather than Doxygen-parsed blocks.  It is also not available from Python.
+ *
+ * With the exception of addMaskPlane, all MaskDict methods provide strong
+ * exception safety.
+ */
+class MaskDict {
+public:
+
+    typedef MaskPlaneDict::value_type value_type;
+    typedef MaskPlaneDict::const_iterator const_iterator;
+
+    // Return the default MaskDict to be used for new Mask instances.
+    static std::shared_ptr<MaskDict> makeMaskDict();
+
+    // Return a new MaskDict with the same plane definitions as the given MaskPlaneDict.
+    static std::shared_ptr<MaskDict> makeMaskDict(detail::MaskPlaneDict const& dict);
+
+    // Set the default MaskDict.
+    static std::shared_ptr<MaskDict> setDefaultDict(std::shared_ptr<MaskDict> dict);
+
+    /*
+     * Set the default MaskDict to a copy of the current one, returning the
+     * new default.
+     */
+    static std::shared_ptr<MaskDict> detachDefault();
+
+    /*
+     * Add the given mask plane to all active MaskDicts for which there is
+     * no conflict, including the default MaskDict.
+     *
+     * MaskDicts that already have a plane with the same name or bit ID
+     * are not affected.
+     *
+     * Provides basic exception safety: mask planes may be added to some
+     * MaskDict instances even if an exception is raised while adding
+     * the mask plane to later ones (though the only exception that could
+     * be thrown in practice is std::bad_alloc).
+     */
+    static void addMaskPlane(std::string const & name, int bitId);
+
+    // Assignment is disabled; we don't need it.
+    MaskDict & operator=(MaskDict const &) = delete;
+    MaskDict & operator=(MaskDict &&) = delete;
+
+    /*
+     * Destructor is potentially non-trivial in order to maintain the global
+     * list of active instances, so even if it is trivial it's best to hide
+     * that in the .cc file.
+     */
+    ~MaskDict();
+
+    // Return a deep copy of the MaskDict.
+    std::shared_ptr<MaskDict> clone() const;
+
+    /*
+     * Return an integer bit ID that is not currently used in this MaskDict.
+     *
+     * No other guarantees are made about the ID.
+     */
+    int getUnusedPlane() const;
```

Comment:
I wasn't sure until you asked - looks like the answer is that it will always return an ID, but there's no guarantee any particular Mask has a pixel type with enough bits for it.  But `MaskDict` doesn't know enough to catch that, so I think the best I can do is document that behavior here.